File: perception/load_env_file.py
"""perception/load_env_file.py -- read `.env` at the repo root into os.environ.

Why this exists: `python -m perception.server` starts a bare interpreter, so a key
sitting in `.env` never reached it. `stages/read_regions.py`'s `load_region_reader`
checks `os.environ` and silently falls back to PaddleOCR (and from there to "no
reader at all") -- correct behaviour when there is genuinely no key, and completely
invisible when there is one sitting on disk. The result, measured on this machine:
`/perceive` reported "PaddleOCR is unavailable; regions were detected but not read"
on every wireframe even with a real, working key in `.env`, because nothing ever
read the file into this process's environment.

This is the same gap `server/src/loadEnvFile.js` closes on the Node side (T-151,
B-012's ".env never loaded" story) -- one process down. The two loaders are
deliberately kept in step: same parsing rules, same non-destructive merge, same
"missing file is not an error" behaviour, so the two languages describe one fact
about the repository rather than two slightly different ones.

Deliberately dependency-free: `python-dotenv` is not installed in `perception/.venv`
(section 12's CPU-only, no-extra-dependency constraint), and this parser is a dozen
lines. Deliberately non-destructive: a variable already present in the real
environment always wins, so a shell export or a test harness overrides the file
rather than fighting it.

AGENTS.md rule 5 is unaffected. A missing file is not an error, and a missing key
still lands on the deterministic path (PaddleOCR, or no reader) exactly as before.
"""

from __future__ import annotations

import os
import pathlib

DEFAULT_ENV_PATH = pathlib.Path(__file__).resolve().parent.parent / ".env"


def parse_env(text: str) -> dict[str, str]:
    """Parse `.env` text. Returns a plain dict; never raises on a malformed line."""
    out: dict[str, str] = {}
    for raw_line in text.splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue
        eq = line.find("=")
        if eq < 1:
            continue
        key = line[:eq].strip()
        if key.startswith("export "):
            key = key[len("export "):].strip()
        if not key or not (key[0].isalpha() or key[0] == "_"):
            continue
        if not all(c.isalnum() or c == "_" for c in key):
            continue
        value = line[eq + 1:].strip()
        if len(value) >= 2 and value[0] in "\"'" and value[-1] == value[0]:
            value = value[1:-1]
        out[key] = value
    return out


def load_env_file(
    env_path: pathlib.Path | str = DEFAULT_ENV_PATH,
    env: "os._Environ[str] | dict[str, str]" = os.environ,
) -> list[str]:
    """Load `.env` into `env` (defaults to `os.environ`) without overwriting anything
    already set. Returns the names of the keys it applied -- names only, so a caller
    can log what was loaded without logging a secret.
    """
    try:
        text = pathlib.Path(env_path).read_text(encoding="utf-8")
    except OSError:
        return []  # no file is the normal case, not a failure

    applied = []
    for key, value in parse_env(text).items():
        if not env.get(key):
            env[key] = value
            applied.append(key)
    return applied

File: server/src/loadEnvFile.js
// server/src/loadEnvFile.js — read `.env` at the repo root into process.env.
//
// Why this exists: `npm run server` starts a bare Node process, so a key sitting in
// `.env` never reached it. Every hosted path (promptToIrHosted, the VLM region reader)
// checks process.env and silently fell back to the deterministic path — correct
// behaviour when there is genuinely no key, and completely invisible when there is one
// sitting on disk. The result was that the API always emitted the reference template
// while the identical call made in-process with the key produced a real, varied IR.
//
// Deliberately dependency-free: `dotenv` is installed in neither package, and a
// twenty-line parser is cheaper than a new dependency. Deliberately non-destructive: a
// variable already present in the real environment always wins, so a shell export, CI
// and the test harness override the file rather than fight it.
//
// AGENTS.md rule 5 is unaffected. A missing file is not an error, and a missing key
// still lands on the deterministic path exactly as before.

import fs from 'node:fs';
import path from 'node:path';
import url from 'node:url';

const HERE = path.dirname(url.fileURLToPath(import.meta.url));
const DEFAULT_ENV_PATH = path.resolve(HERE, '..', '..', '.env');

/** Parse `.env` text. Returns a plain object; never throws on a malformed line. */
export function parseEnv(text) {
  const out = {};
  for (const rawLine of String(text).split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line || line.startsWith('#')) continue;
    const eq = line.indexOf('=');
    if (eq < 1) continue;
    const key = line.slice(0, eq).trim().replace(/^export\s+/, '');
    if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(key)) continue;
    let value = line.slice(eq + 1).trim();
    // Strip one matched pair of surrounding quotes, if present.
    if (value.length >= 2 && (value[0] === '"' || value[0] === "'") && value[value.length - 1] === value[0]) {
      value = value.slice(1, -1);
    }
    out[key] = value;
  }
  return out;
}

/**
 * Load `.env` into `env` (defaults to process.env) without overwriting anything already
 * set. Returns the names of the keys it applied — names only, so a caller can log what
 * was loaded without logging a secret.
 */
export function loadEnvFile({ envPath = DEFAULT_ENV_PATH, env = process.env } = {}) {
  let text;
  try {
    text = fs.readFileSync(envPath, 'utf8');
  } catch {
    return [];               // no file is the normal case, not a failure
  }
  const applied = [];
  for (const [key, value] of Object.entries(parseEnv(text))) {
    if (env[key] === undefined || env[key] === '') {
      env[key] = value;
      applied.push(key);
    }
  }
  return applied;
}

export default loadEnvFile;
